### ISSUE TITLE
Error: failed to execute "wasm-pack" "build"

### ISSUE THREAD

**Issue by expilo** [tinysearch/tinysearch#151]
When trying to build nightly with `docker build --build-arg RUST_IMAGE=rustlang/rust:nightly-alpine -t tinysearch/cli:nightly .` I'm getting the following error:
```
thread 'main' panicked at 'crate directory should exist', src/readme.rs:11:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Well, this is embarrassing.

wasm-pack had a problem and crashed. To help us diagnose the problem you can send us a crash report.


We have generated a report file at "/tmp/report-2ec64b30-c702-4581-a790-5d71fb2f66fe.toml". Submit an issue or email with the subject of "wasm-pack Crash Report" and include the report as an attachment.

- Authors: Ashley Williams <ashley666ashley@gmail.com>, Jesper Håkansson <jesper@jesperh.se>

We take privacy seriously, and do not perform any automated error collection. In order to improve the software, we rely on people to submit reports.

Thank you kindly!
Error: failed to execute "wasm-pack" "build" "/tmp/.tmpzklwBA/engine" "--target" "web" "--release" "--out-dir" "/tmp"
status: exit status: 101
```
In spite of the message no report file was generated at `tmp/report-2ec64b30-c702-4581-a790-5d71fb2f66fe.toml`

**Comment by mre** [comment]
I think this is another variant of https://github.com/tinysearch/tinysearch/issues/147, which I thought I fixed with 0.6.3, but I didn't. Still trying to figure out what's going on.

**Comment by expilo** [comment]
You made me start learning rust, which I did not plan :) It seems tinysearch fails on the line
`fs::rename("storage", engine_dir.join("storage"))?;`

**Comment by expilo** [comment]
OK, I guess I have found the reason:

```
pub fn rename<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> Result<()>
Rename a file or directory to a new name, replacing the original file if to already exists.
This will not work if the new name is on a different mount point.
```

Indeed, my "/tmp" directory is on a different mount point.

**Comment by expilo** [comment]
So I have to run tinysearch on the same partition that `/tmp` is on. Then I can compile index.json fine.

**Comment by mre** [comment]
I created a pull request here: https://github.com/tinysearch/tinysearch/pull/154.
If someone could test this and report back I'd be very thankful.

**Comment by amousset** [comment]
On 0.6.4 I get:

```
Unpacking tinysearch WASM engine into temporary directory "/tmp/.tmpJuEao6"
Starting unpack
Copying index into crate
Compiling WASM module using wasm-pack
Error: crate directory is missing a `Cargo.toml` file; is `/tmp/.tmpJuEao6/engine` the wrong directory?
Error: failed to execute "wasm-pack" "build" "/tmp/.tmpJuEao6/engine" "--target" "web" "--release" "--out-dir" "/home/amousset/projects/rustsec/admin/wasm_output"
status: exit status: 1
```

which is not exactly the same error but looks close.

**Comment by ms-ati** [comment]
I have the same error as just above ☝🏻 also using 0.6.4

**Comment by davidlampach** [comment]
I am using the latest wasm-pack and I am getting this error exactly EVERY OTHER time I compile.  Has anyone figured it out?

**Comment by Jieiku** [comment]
I think this may no longer be an issue?

I did an install today and have no problem:

```bash
sudo pacman -S install wasm-pack binaryen
git clone https://github.com/tinysearch/tinysearch
cd tinysearch
cargo build --release
sudo cp ./target/release/tinysearch /usr/local/bin/tinysearch
exit
```

```bash
tinysearch --optimize --path static public/data_tinysearch/index.html
```

output:

```
Unpacking tinysearch WASM engine into temporary directory "/tmp/.tmpZ3Y4lx"
Starting unpack
Copying "../engine/src/lib.rs"
Copying "../engine/Cargo.toml"
Copying "../shared/.gitignore"
Copying "../engine/.gitignore"
Copying "../shared/src/lib.rs"
Copying "../shared/Cargo.toml"
Copying index into crate
Compiling WASM module using wasm-pack
[INFO]: Checking for the Wasm target...
[INFO]: Compiling to Wasm...
   Compiling proc-macro2 v1.0.56
   Compiling unicode-ident v1.0.8
   Compiling quote v1.0.26
   Compiling serde_derive v1.0.159
   Compiling serde v1.0.159
   Compiling wasm-bindgen-shared v0.2.84
   Compiling syn v1.0.109
   Compiling log v0.4.17
   Compiling cfg-if v1.0.0
   Compiling bumpalo v3.12.0
   Compiling once_cell v1.17.1
   Compiling serde_json v1.0.95
   Compiling ryu v1.0.13
   Compiling itoa v1.0.6
   Compiling wasm-bindgen v0.2.84
   Compiling wee_alloc v0.4.5
   Compiling memory_units v0.4.0
   Compiling cfg-if v0.1.10
   Compiling syn v2.0.13
   Compiling wasm-bindgen-backend v0.2.84
   Compiling wasm-bindgen-macro-support v0.2.84
   Compiling wasm-bindgen-macro v0.2.84
   Compiling xorf v0.7.2
   Compiling bincode v1.3.3
   Compiling tinysearch-shared v0.7.0 (/tmp/.tmpZ3Y4lx/shared)
   Compiling tinysearch-engine v0.7.0 (/tmp/.tmpZ3Y4lx/engine)
warning: use of deprecated associated function `wasm_bindgen::JsValue::from_serde`: causes dependency cycles, use `serde-wasm-bindgen` or `gloo_utils::format::JsValueSerdeExt` instead
  --> src/lib.rs:54:14
   |
54 |     JsValue::from_serde(&results).unwrap()
   |              ^^^^^^^^^^
   |
   = note: `#[warn(deprecated)]` on by default

warning: `tinysearch-engine` (lib) generated 1 warning
    Finished release [optimized] target(s) in 11.93s
[WARN]: :-) origin crate has no README
[INFO]: License key is set in Cargo.toml but no LICENSE file(s) were found; Please add the LICENSE file(s) to your project directory
[INFO]: Installing wasm-bindgen...
[INFO]: found wasm-opt at "/usr/bin/wasm-opt"
[INFO]: Optimizing wasm binaries with `wasm-opt`...
[INFO]: Optional field missing from Cargo.toml: 'repository'. This is not necessary, but recommended
[INFO]: :-) Done in 12.41s
[INFO]: :-) Your wasm pkg is ready to publish at /home/jieiku/.dev/abridge/static.
[WARN]: :-) There's a newer version of wasm-pack available, the new version is: 0.11.0, you are using: 0.10.3. To update, navigate to: https://rustwasm.github.io/wasm-pack/installer/
All done! Open the output folder with a web server to try the demo.
```

**Comment by mre** [comment]
Guess it's still a problem when you directly `cargo install tinysearch` as the files get stripped and are missing during installation. That's what I understood at least.

**Comment by Jieiku** [comment]
On a Fresh Ubuntu 22.04 VM I got this far when trying to install tinysearch via cargo:

```bash
sudo apt install build-essential
sudo apt install make gcc cmake binaryen curl

curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y

sudo reboot

cargo install wasm-pack

cargo install tinysearch

wget https://raw.githubusercontent.com/tinysearch/tinysearch/master/fixtures/index.json

tinysearch --optimize index.json
```

```
Unpacking tinysearch WASM engine into temporary directory "/tmp/.tmpcrGNXQ"
Starting unpack
Copying index into crate
Compiling WASM module using wasm-pack
Error: crate directory is missing a `Cargo.toml` file; is `/tmp/.tmpcrGNXQ/engine` the wrong directory?
Caused by: crate directory is missing a `Cargo.toml` file; is `/tmp/.tmpcrGNXQ/engine` the wrong directory?
Error: failed to execute "wasm-pack" "build" "/tmp/.tmpcrGNXQ/engine" "--target" "web" "--release" "--out-dir" "/home/jake/wasm_output"
status: exit status: 1
```

If there is anything else I should check or try let me know :)

**Comment by mre** [comment]
Can you run the test again with this branch?
https://github.com/tinysearch/tinysearch/pull/168

If that works, can you publish tinysearch from this branch on crates.io, e.g. as `tinysearch-tmp` and then test again with `cargo install tinysearch-tmp`?
It's a big ask, but it would help a lot to unblock the project.
I'm inclined to merge that PR if it fixes the issue, even if I don't have the time to fix the failing tests right now.

**Comment by Jieiku** [comment]
yes, I will give it a try in a couple hours.

**Comment by Jieiku** [comment]
Fresh Ubuntu 22.04 VM:

```bash
sudo apt install build-essential
sudo apt install make gcc cmake binaryen curl git
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
sudo reboot

cargo install wasm-pack

git clone --branch fix-for-flex https://github.com/ASMfreaK/tinysearch
cd tinysearch
cargo build --release
```

```
   Compiling proc-macro2 v1.0.51
   Compiling unicode-ident v1.0.7
   Compiling syn v1.0.109
   Compiling serde_derive v1.0.127
   Compiling quote v1.0.9
   Compiling serde v1.0.127
   Compiling bincode v1.3.3
   Compiling xorf v0.7.2
   Compiling tinysearch v0.7.0 (/home/jake/tinysearch/tinysearch)
    Finished release [optimized] target(s) in 13.69s
```

Normally this would result in the executable being created: `./target/release/tinysearch` but I am not seeing it.

Does this new branch maybe require more steps to build the final tinysearch executable?

here is a tree view: of what the build seems to have produced:


```
drwxrwxr-x    - jake jake 13 Apr 20:01  tinysearch
drwxrwxr-x    - jake jake 13 Apr 20:01 ├──  assets
drwxrwxr-x    - jake jake 13 Apr 20:01 │  ├──  crate
.rw-rw-r--  917 jake jake 13 Apr 20:01 │  │  ├──  Cargo.toml
drwxrwxr-x    - jake jake 13 Apr 20:01 │  │  └──  src
.rw-rw-r--  765 jake jake 13 Apr 20:01 │  │     └──  lib.rs
.rw-rw-r-- 2.3k jake jake 13 Apr 20:01 │  ├──  demo.html
.rw-rw-r-- 4.9k jake jake 13 Apr 20:01 │  └──  stopwords
.rw-rw-r-- 1.3k jake jake 13 Apr 20:01 ├──  Cargo.toml
.rw-rw-r--  117 jake jake 13 Apr 20:01 ├──  .gitignore
drwxrwxr-x    - jake jake 13 Apr 20:01 └──  src
drwxrwxr-x    - jake jake 13 Apr 20:01    ├──  bin
.rw-rw-r--  13k jake jake 13 Apr 20:01    │  ├──  tinysearch.rs
drwxrwxr-x    - jake jake 13 Apr 20:01    │  └──  utils
.rw-rw-r--  514 jake jake 13 Apr 20:01    │     ├──  assets.rs
.rw-rw-r--  329 jake jake 13 Apr 20:01    │     ├──  index.rs
.rw-rw-r--   48 jake jake 13 Apr 20:01    │     ├──  mod.rs
.rw-rw-r-- 4.4k jake jake 13 Apr 20:01    │     └──  storage.rs
.rw-rw-r-- 2.5k jake jake 13 Apr 20:01    └──  lib.rs
```

**Comment by mre** [comment]
@ASMfreaK combined the bin and library in the same crate in their branch.
You need
```
cargo build --release --features=bin
```

to build the binary.

**Comment by Jieiku** [comment]
OK, I got the executable built, but the executable no longer works.
This might fix the Crate, but it seems it causes problems when you build from git?

```bash
sudo apt install build-essential
sudo apt install make gcc cmake binaryen curl git
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
sudo reboot

cargo install wasm-pack

git clone --branch fix-for-flex https://github.com/ASMfreaK/tinysearch
cd tinysearch
cargo build --release
cargo build --release --features=bin
cd
wget https://raw.githubusercontent.com/tinysearch/tinysearch/master/fixtures/index.json
~/tinysearch/target/release/tinysearch --optimize index.json
```

```
Creating tinysearch implementation crate tinysearch-engine in directory /tmp/.tmpKjWsQQ
Creating storage file for posts index.json in file /tmp/.tmpKjWsQQ/src/storage
Error: Failed to build wasm mode

Caused by:
    0: Failed generating crate
    1: Failed building storage
    2: No such file or directory (os error 2)
```

**Comment by ASMfreaK** [comment]
@Jieiku For the moment, my branch requires using this option: `-e, --engine-version`.

So, to refactor your script, you would need to run:
```
cd
wget https://raw.githubusercontent.com/tinysearch/tinysearch/master/fixtures/index.json
~/tinysearch/target/release/tinysearch -e "path=\"$HOME/tinysearch/tinysearch\"" --optimize index.json
```

**Comment by Jieiku** [comment]
Thanks! I got farther:

```
~/tinysearch/target/release/tinysearch -e "path=\"$HOME/tinysearch/tinysearch\"" --optimize index.json
Creating tinysearch implementation crate tinysearch-engine in directory /tmp/.tmpBa4BEy
Creating storage file for posts index.json in file /tmp/.tmpBa4BEy/src/storage
Storage ready in file /tmp/.tmpBa4BEy/src/storage
Crate content generated in /tmp/.tmpBa4BEy/
Compiling WASM module using wasm-pack
running "wasm-pack" "build" "/tmp/.tmpBa4BEy" "--target" "web" "--release" "--out-dir" "/home/jake/wasm_output"
[INFO]: 🎯  Checking for the Wasm target...
info: downloading component 'rust-std' for 'wasm32-unknown-unknown'
info: installing component 'rust-std' for 'wasm32-unknown-unknown'
[INFO]: 🌀  Compiling to Wasm...
   Compiling proc-macro2 v1.0.56
   Compiling unicode-ident v1.0.8
   Compiling quote v1.0.26
   Compiling log v0.4.17
   Compiling syn v1.0.109
   Compiling wasm-bindgen-shared v0.2.84
   Compiling serde_derive v1.0.160
   Compiling cfg-if v1.0.0
   Compiling syn v2.0.15
   Compiling bumpalo v3.12.0
   Compiling once_cell v1.17.1
   Compiling serde v1.0.160
   Compiling wasm-bindgen-backend v0.2.84
   Compiling wasm-bindgen-macro-support v0.2.84
   Compiling wasm-bindgen v0.2.84
   Compiling wasm-bindgen-macro v0.2.84
   Compiling wee_alloc v0.4.5
   Compiling js-sys v0.3.61
   Compiling xorf v0.7.2
   Compiling bincode v1.3.3
   Compiling memory_units v0.4.0
   Compiling cfg-if v0.1.10
   Compiling tinysearch v0.7.0 (/home/jake/tinysearch/tinysearch)
   Compiling serde-wasm-bindgen v0.4.5
   Compiling tinysearch-engine v0.7.0 (/tmp/.tmpBa4BEy)
    Finished release [optimized] target(s) in 21.97s
[WARN]: ⚠   origin crate has no README
[INFO]: License key is set in Cargo.toml but no LICENSE file(s) were found; Please add the LICENSE file(s) to your project directory
[INFO]: ⬇  Installing wasm-bindgen...
[INFO]: found wasm-opt at "/usr/bin/wasm-opt"
Error: /usr/bin/bin/wasm-opt binary does not exist
To disable `wasm-opt`, add `wasm-opt = false` to your package metadata in your `Cargo.toml`.
Caused by: /usr/bin/bin/wasm-opt binary does not exist
To disable `wasm-opt`, add `wasm-opt = false` to your package metadata in your `Cargo.toml`.
Error: Failed to build wasm mode

Caused by:
    failed to execute "wasm-pack" "build" "/tmp/.tmpBa4BEy" "--target" "web" "--release" "--out-dir" "/home/jake/wasm_output"
    status: exit status: 1
```

**Comment by ASMfreaK** [comment]
It works fo me with `cargo run --force wasm-pack --vers=^0.9`, but we should fix that error

**Comment by Jieiku** [comment]
AHH I did not know you can force a version like that. ( I am still very new to rust )

So that confirms the problem is with the 0.11.0 version, now just have to figure out why it happens.

**Comment by mre** [comment]
Closing this issue as resolved as of 0.8.2. At least I can't reproduce the error anymore.

The reason why it broke was a combination of things. One major one was a limitation by cargo: it does not allow including one crate inside another one. This is done to avoid unwanted complexity and is a sane default. In our case, however, it leads to the situation that the engine crate was missing from the tinysearch crate on crates.io.
See also https://github.com/tinysearch/tinysearch/issues/170#issuecomment-1650709333.

The latest tinysearch 0.8.2 fixes this issue. Apologies for the long delay to produce a fix.

**Comment by millette** [comment]
I commented <https://github.com/tinysearch/tinysearch/issues/170#issuecomment-1654656395>.
